packages/nx/src/command-line/init/configure-plugins.ts

```diff
@@ -0,0 +1,159 @@
+import * as createSpinner from 'ora';
+import { bold } from 'chalk';
+
+import {
+  getPackageManagerCommand,
+  PackageManagerCommands,
+} from '../../utils/package-manager';
+import { GitRepository } from '../../utils/git-utils';
+import { output } from '../../utils/output';
+import { flushChanges, FsTree } from '../../generators/tree';
+import {
+  Generator as NxGenerator,
+  GeneratorCallback,
+} from '../../config/misc-interfaces';
+import { getGeneratorInformation } from '../generate/generator-utils';
+import { workspaceRoot } from '../../utils/workspace-root';
+import { addDepsToPackageJson, runInstall } from './implementation/utils';
+
+export function runPackageManagerInstallPlugins(
+  repoRoot: string,
+  pmc: PackageManagerCommands = getPackageManagerCommand(),
+  plugins: string[]
+) {
+  if (plugins.length === 0) {
+    return;
+  }
+  addDepsToPackageJson(repoRoot, plugins);
+  runInstall(repoRoot, pmc);
+}
+
+/**
+ * Install plugins
+ * Get the implementation of the plugin's init generator and run it
+ * @returns a list of succeeded plugins and a map of failed plugins to errors
+ */
+export async function installPlugins(
+  plugins: string[],
+  updatePackageScripts: boolean,
+  repoRoot: string = workspaceRoot,
+  verbose: boolean = false
+): Promise<{
+  succeededPlugins: string[];
+  failedPlugins: { [plugin: string]: Error };
+}> {
+  if (plugins.length === 0) {
+    return {
+      succeededPlugins: [],
+      failedPlugins: {},
+    };
+  }
+  output.log({ title: '🔨 Configuring plugins' });
+  const spinner = createSpinner();
+  let succeededPlugins = [];
+  const failedPlugins: {
+    [pluginName: string]: Error;
+  } = {};
+  const host = new FsTree(
+    repoRoot,
+    verbose,
+    `install plugins ${plugins.join(' ')}`
+  );
+  for (const plugin of plugins) {
+    spinner.start('Installing plugin ' + plugin);
+    try {
+      const { implementationFactory } = getGeneratorInformation(
+        plugin,
+        'init',
+        repoRoot,
+        {}
+      );
+      const implementation: NxGenerator = implementationFactory();
+      const task: GeneratorCallback | void = await implementation(host, {
+        keepExistingVersions: true,
+        updatePackageScripts,
+      });
+      if (task) {
+        await task();
+      }
+      succeededPlugins.push(plugin);
+      spinner.succeed('Installed plugin ' + plugin);
+    } catch (e) {
+      failedPlugins[plugin] = e;
+      spinner.fail('Failed to install plugin ' + plugin);
+    }
+  }
+  host.lock();
+  const changes = host.listChanges();
+  flushChanges(repoRoot, changes);
```

Comment:
Ideally, this would be able to work like this but the host will have partial changes when failures are encountered.

A new tree should be created and flushed before the next plugin starts and creates a new tree much like it does currently.